This is RFAL.

An example corruption-like test failure if we do use the vectorized encoder with concurrent ALTER PK:
```
panic: invalid header size: 4
  (1) attached stack trace
    -- stack trace:
    | runtime.gopanic
    |   GOROOT/src/runtime/panic.go:783
    | [...repeated from below...]
  Wraps: (2) attached stack trace
    -- stack trace:
    | github.com/cockroachdb/cockroach/pkg/roachpb.Value.VerifyHeader
    |   pkg/roachpb/data.go:383
    | github.com/cockroachdb/cockroach/pkg/kv.marshalValue
    |   pkg/kv/util.go:47
    | github.com/cockroachdb/cockroach/pkg/kv.(*Batch).put
    |   pkg/kv/batch.go:437
    | github.com/cockroachdb/cockroach/pkg/kv.(*Batch).Put
    |   pkg/kv/batch.go:466
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).constructMergeBatch
    |   pkg/sql/backfill/mvcc_index_merger.go:599
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).merge.func1.1
    |   pkg/sql/backfill/mvcc_index_merger.go:480
    | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn.func1
    |   pkg/sql/internal.go:1965
    | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func4
    |   pkg/sql/internal.go:2052
    | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec
    |   pkg/kv/txn.go:1144
    | github.com/cockroachdb/cockroach/pkg/kv.runTxn
    |   pkg/kv/db.go:1101
    | github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl
    |   pkg/kv/db.go:1064
    | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn.func2
    |   pkg/sql/internal.go:2025
    | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).txn
    |   pkg/sql/internal.go:2039
    | github.com/cockroachdb/cockroach/pkg/sql.(*InternalDB).Txn
    |   pkg/sql/internal.go:1966
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).merge.func1
    |   pkg/sql/backfill/mvcc_index_merger.go:418
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.retryWithReducedBatchWhenAutoRetryLimitExceeded
    |   pkg/sql/backfill/mvcc_index_merger.go:719
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).merge
    |   pkg/sql/backfill/mvcc_index_merger.go:416
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).Run.func5
    |   pkg/sql/backfill/mvcc_index_merger.go:226
    | github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfillMerger).Run.Group.GoCtx.func8
    |   pkg/util/ctxgroup/ctxgroup.go:200
    | golang.org/x/sync/errgroup.(*Group).Go.func1
    |   external/org_golang_x_sync/errgroup/errgroup.go:93
    | runtime.goexit
    |   src/runtime/asm_amd64.s:1693
```
I spent little time looking into this but not sure what was going wrong so decided to not use the vectorized engine for COPY when ForcePut is present but still add the support for it.

--- a/pkg/sql/colenc/encode.go
+++ b/pkg/sql/colenc/encode.go
@@ -125,11 +125,6 @@ func (b *BatchEncoder) PrepareBatch(ctx context.Context, p row.Putter, start, en
 	}
 	for _, ind := range b.rh.TableDesc.WritableNonPrimaryIndexes() {
 		b.resetBuffers()
-		// TODO(cucaroach): COPY doesn't need ForcePut support but the encoder
-		// will need to support it eventually.
-		if ind.ForcePut() {
-			colexecerror.InternalError(errors.AssertionFailedf("vector encoder doesn't support ForcePut yet"))
-		}
 		if err := b.encodeSecondaryIndex(ctx, ind); err != nil {
 			return err
 		}
@@ -484,6 +479,13 @@ func (b *BatchEncoder) encodeSecondaryIndex(ctx context.Context, ind catalog.Ind
 	return b.checkMemory()
 }
 
+func (b *BatchEncoder) useCPutForSecondary(ind catalog.Index) bool {
+	if ind.ForcePut() {
+		return false
+	}
+	return ind.IsUnique() || b.useCPutsOnNonUniqueIndexes
+}
+
 func (b *BatchEncoder) encodeSecondaryIndexNoFamilies(ind catalog.Index, kys []roachpb.Key) error {
 	for row := 0; row < b.count; row++ {
 		// Elided partial index keys will be empty.
@@ -510,7 +512,7 @@ func (b *BatchEncoder) encodeSecondaryIndexNoFamilies(ind catalog.Index, kys []r
 	if err := b.writeColumnValues(kys, values, ind, cols); err != nil {
 		return err
 	}
-	if ind.IsUnique() || b.useCPutsOnNonUniqueIndexes {
+	if b.useCPutForSecondary(ind) {
 		b.p.CPutBytesEmpty(kys, values)
 	} else {
 		b.p.PutBytes(kys, values)
@@ -576,13 +578,13 @@ func (b *BatchEncoder) encodeSecondaryIndexWithFamilies(
 		// include encoded primary key columns. For other families,
 		// use the tuple encoding for the value.
 		if familyID == 0 {
-			if ind.IsUnique() || b.useCPutsOnNonUniqueIndexes {
+			if b.useCPutForSecondary(ind) {
 				b.p.CPutBytesEmpty(kys, values)
 			} else {
 				b.p.PutBytes(kys, values)
 			}
 		} else {
-			if ind.IsUnique() || b.useCPutsOnNonUniqueIndexes {
+			if b.useCPutForSecondary(ind) {
 				b.p.CPutTuplesEmpty(kys, values)
 			} else {
 				b.p.PutTuples(kys, values)

--- a/pkg/sql/colenc/inverted.go
+++ b/pkg/sql/colenc/inverted.go
@@ -94,7 +94,7 @@ func (b *BatchEncoder) encodeInvertedSecondaryIndexNoFamiliesOneRow(
 	}
 	var kvValue roachpb.Value
 	kvValue.SetBytes(value)
-	if ind.IsUnique() || b.useCPutsOnNonUniqueIndexes {
+	if b.useCPutForSecondary(ind) {
 		b.p.CPut(&key, &kvValue, nil /* expValue */)
 	} else {
 		b.p.Put(&key, &kvValue)

--- a/pkg/sql/copy/copy_test.go
+++ b/pkg/sql/copy/copy_test.go
@@ -704,8 +704,9 @@ func TestLargeCopy(t *testing.T) {
 	// This test can cause timeouts.
 	skip.UnderRace(t)
 	ctx := context.Background()
+	rng, _ := randutil.NewPseudoRand()
 
-	srv, _, kvdb := serverutils.StartServer(t, base.TestServerArgs{})
+	srv, sqlDB, kvdb := serverutils.StartServer(t, base.TestServerArgs{})
 	defer srv.Stopper().Stop(ctx)
 	s := srv.ApplicationLayer()
 
@@ -719,16 +720,53 @@ func TestLargeCopy(t *testing.T) {
 
 	desc := desctestutils.TestingGetPublicTableDescriptor(kvdb, s.Codec(), "defaultdb", "lineitem")
 	require.NotNil(t, desc, "Failed to lookup descriptor")
+	rows := rng.Intn(2000) + 1
+	cr := &copyReader{rng: rng, cols: desc.PublicColumns(), rows: rows}
 
-	err = conn.Exec(ctx, "SET copy_from_atomic_enabled = false")
-	require.NoError(t, err)
+	if rng.Float64() < 0.5 {
+		_, err = sqlDB.Exec("SET copy_from_atomic_enabled = false")
+		require.NoError(t, err)
+	}
 
-	rng := rand.New(rand.NewSource(0))
-	rows := 100
-	numrows, err := conn.GetDriverConn().CopyFrom(ctx,
-		&copyReader{rng: rng, cols: desc.PublicColumns(), rows: rows},
-		"COPY lineitem FROM STDIN WITH CSV;")
-	require.NoError(t, err)
+	// In 50% cases change the primary key concurrently with the COPY.
+	doAlterPK := rng.Float64() < 0.5
+	ignoreErr := func(err error) bool {
+		if err == nil {
+			return true
+		}
+		if !doAlterPK {
+			return false
+		}
+		// We might hit a duplicate key error when changing the primary key
+		// (which seems somewhat expected), so we'll ignore such an error.
+		return strings.Contains(err.Error(), "duplicate key") ||
+			// TODO(yuzefovich): occasionally we get this error, and it's not
+			// clear why. Look into this.
+			strings.Contains(err.Error(), "cannot disable pipelining on a running transaction")
+	}
+	alterPKCh := make(chan error)
+	if doAlterPK {
+		// We'll delay starting the ALTER to exercise different scenarios.
+		sleepMillis := rng.Intn(2000)
+		go func() {
+			defer close(alterPKCh)
+			time.Sleep(time.Duration(sleepMillis) * time.Millisecond)
+			_, err := sqlDB.Exec("ALTER TABLE lineitem ALTER PRIMARY KEY USING COLUMNS (l_orderkey, l_linenumber, l_suppkey);")
+			alterPKCh <- err
+		}()
+		// Delay starting the COPY too.
+		time.Sleep(time.Duration(rng.Intn(2000)) * time.Millisecond)
+	} else {
+		close(alterPKCh)
+	}
+	numrows, copyErr := conn.GetDriverConn().CopyFrom(ctx, cr, "COPY lineitem FROM STDIN WITH CSV;")
+	alterErr := <-alterPKCh
+	if !ignoreErr(copyErr) {
+		t.Fatal(copyErr)
+	}
+	if !ignoreErr(alterErr) {
+		t.Fatal(alterErr)
+	}
 	require.Equal(t, int(numrows), rows)
 }
 

--- a/pkg/sql/copy_from.go
+++ b/pkg/sql/copy_from.go
@@ -428,6 +428,19 @@ func (c *copyMachine) canSupportVectorized(table catalog.TableDescriptor) bool {
 	if len(table.VectorIndexes()) > 0 {
 		return false
 	}
+	forcePut := table.GetPrimaryIndex().ForcePut()
+	secondaryIndexes := table.WritableNonPrimaryIndexes()
+	for i := 0; !forcePut && i < len(secondaryIndexes); i++ {
+		forcePut = secondaryIndexes[i].ForcePut()
+	}
+	if forcePut {
+		// Even though the vector encoder supports ForcePut behavior, testing
+		// COPY with a concurrent ALTER PRIMARY KEY has resulted in different
+		// corruption scenarios. The non-vectorized COPY doesn't hit those, so
+		// we choose to fall back.
+		// TODO(#157198): investigate this.
+		return false
+	}
 	// Vectorized COPY doesn't support foreign key checks, no reason it couldn't
 	// but it doesn't work right now because we don't have the ability to
 	// hold the results in a bufferNode. We wouldn't want to enable it

